pkg/volume/aws_ebs/aws_util.go

```diff
@@ -112,62 +148,76 @@ func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner, node *
 			}
 		case "kmskeyid":
 			volumeOptions.KmsKeyId = v
-		case volume.VolumeParameterFSType:
-			fstype = v
 		default:
-			return "", 0, nil, "", fmt.Errorf("invalid option %q for volume plugin %s", k, c.plugin.GetPluginName())
+			return nil, fmt.Errorf("invalid option %q for volume plugin %s", k, pluginName)
 		}
 	}
 
 	if volumeOptions.ZonePresent && volumeOptions.ZonesPresent {
-		return "", 0, nil, "", fmt.Errorf("both zone and zones StorageClass parameters must not be used at the same time")
+		return nil, fmt.Errorf("both zone and zones StorageClass parameters must not be used at the same time")
 	}
 
-	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) && node != nil {
-		if node.ObjectMeta.Labels == nil {
-			return "", 0, nil, "", fmt.Errorf("labels for node missing")
-		}
-		z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
-		if !ok {
-			return "", 0, nil, "", fmt.Errorf("%s Label for node missing", kubeletapis.LabelZoneFailureDomain)
-		}
-
-		if volumeOptions.ZonePresent || volumeOptions.ZonesPresent {
-			return "", 0, nil, "", fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode set to WaitForFirstConsumer.")
-		}
-
-		volumeOptions.ZonePresent = true
-		volumeOptions.AvailabilityZone = z
-	}
-
-	// TODO: implement PVC.Selector parsing
-	if c.options.PVC.Spec.Selector != nil {
-		return "", 0, nil, "", fmt.Errorf("claim.Spec.Selector is not supported for dynamic provisioning on AWS")
-	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) {
+		if node != nil {
+			// if node set, zone/zones cannot be specified as node's zone may conflict
+			if volumeOptions.ZonePresent || volumeOptions.ZonesPresent {
+				return nil, fmt.Errorf("Zone/Zones cannot be specified in StorageClass if VolumeBindingMode set to WaitForFirstConsumer.")
+			}
 
-	name, err := cloud.CreateDisk(volumeOptions)
-	if err != nil {
-		glog.V(2).Infof("Error creating EBS Disk volume: %v", err)
-		return "", 0, nil, "", err
-	}
-	glog.V(2).Infof("Successfully created EBS Disk volume %s", name)
+			// if node is set, we pick node's zone and ignore any allowedTopologies (since scheduler has accounted for it already)
+			if node.ObjectMeta.Labels == nil {
+				return nil, fmt.Errorf("labels for node missing")
+			}
 
-	labels, err := cloud.GetVolumeLabels(name)
-	if err != nil {
-		// We don't really want to leak the volume here...
-		glog.Errorf("error building labels for new EBS volume %q: %v", name, err)
+			z, ok := node.ObjectMeta.Labels[kubeletapis.LabelZoneFailureDomain]
+			if !ok {
+				return nil, fmt.Errorf("%s Label for node missing", kubeletapis.LabelZoneFailureDomain)
+			}
+			volumeOptions.ZonePresent = true
+			volumeOptions.AvailabilityZone = z
+		} else {
+			// node not specified so collect all specified zones from allowedTopologies
+			allowedZones := make([]string, 0)
+			for _, term := range allowedTopologies {
```

Comment:
This could be a volume util that gce pd could use too

Reply:
Actually @verult may have some comments here since he is looking at doing a similar allowedTopologies conversion for CSI

Reply:
I'm not sure if multi-zonal PDs will be different.  This function is just creating a list of allowed zones.  The only difference is that for multi-zonal PDs, it will pick 2 zones from the list instead of 1.

Reply:
If we restrict the allowed set of keys in expressions for `allowedTopology` to only `LabelZoneFailureDomain` for now, it certainly simplifies things. The difference in logic compared to single-zone PD in this case will be: 

+ If both node and `allowedTopology` is present, we need to pick the node's zone + one separate zone from `allowedTopology` that is in the same region as node and return those two. 
+ If no node present, we can pick two zones from `allowedTopology` (as you said) and also need to make sure they are from the same region.

Sidenote: It does not appear if zones from multiple regions are specified in `zones` today, there are any checks [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/util.go#L430) to pick zones only from the same region. If that is the case, I can enhance that.

Reply:
I think we can split this out into two operations:
1. Convert allowed topologies to a list of zones
2. Figuring out how to select zones when node is set or not

Also, the zones are all within the same region.  We do not have cross-regional clusters today

Reply:
@verult we may even want something like:
0. Flatten allowed topologies to a list of maps (csi topology format)
1. Convert list of maps to list of zones (and error if there's more than one key)
2. Figure out how to select zones from node + list of zones

Reply:
> I don't think we'll be able to reuse the logic for CSI conversion as it'll be more complicated, for example there isn't a one-to-one mapping between allowedTopology and CSI topology representation.

Can you give an example of what you mean?  If there is no one-to-one mapping, then how will you be able to convert it?

Reply:
> This could be a volume util that gce pd could use too
Should we pull it out into volume util when we get to the GCE plugin? It shouldn't be too hard to do it later on. It's easy to make the wrong assumptions about commonalities if this is done prematurely. I think this is more of a problem when creating abstraction boundaries so I don't have a strong preference either way for this case.

> 0. Flatten allowed topologies to a list of maps (csi topology format)
I don't think we should complicate the logic in in-tree volume plugins for the sake of aligning with CSI implementation details.

> Can you give an example of what you mean? If there is no one-to-one mapping, then how will you be able to convert it?
AllowedTopologies can have ORs of ANDs of ORs, so there are multiple possible representations of the same topology constraint. CSI only has ORs of ANDs. To convert from CSI representation to AllowedTopologies, we could pick a canonical representation, similar to how we deal with the same issue with volume node affinity.

Reply:
Ok, let's consider refactoring this when we get to the CSI implementation.  But I feel the in-tree plugins use of allowed topologies is a subset of the CSI functionality, so it should be possible to get the list of zones out of the CSI representation.  Plus, here we're only talking about converting from AllowedTopologies -> CSI, not the other way around.

So for now, I think a common util that only handles zones should be fine

Reply:
> CSI only has ORs of ANDs

@verult Why does CSI not support ORs of ANDs of ORs? For multi-zonal implementations like GCE PD, the "last" ORs bit corresponding to the `values` is where the possible set of zones are specified:

```
allowedTopologies:
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-a
    - us-central1-b
    - us-central1-c
```
Will the above convention change for multi-zonal volumes in CSI?

Reply:
AFAIU we could similarly have
```
allowedTopologies:
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-a
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-b
- matchLabelExpressions:
  - key: failure-domain.beta.kubernetes.io/zone
    values:
    - us-central1-c
```
to represent the same thing (in both cases we are saying, these are the zones the volume plugin is free to choose from), and an equivalent form of this is how CSI represents it